Source/WebCore/accessibility/AXTextMarker.cpp

```diff
@@ -933,7 +933,41 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
 
         return resultMarker;
     }
-    // FIXME: Implement the other AXTextUnits.
+
+    // Paragraphs must be handled differently from word + sentence boundaries, as there is no paragraph break iterator.
+    // Rather, paragraph boundaries are based on rendered newlines and changes in nodes/renderers.
+    if (textUnit == AXTextUnit::Paragraph) {
+        size_t startLineIndex = currentRuns->lineID(runIndex).lineIndex;
+        unsigned offsetInStartLine = currentRuns->runLengthSumTo(startLineIndex - 1) - offset();
+
+        while (currentObject) {
+            RELEASE_ASSERT(currentRuns->size());
+            unsigned cumulativeOffset = direction == AXDirection::Next ? 0 : currentRuns->runLengthSumTo(startLineIndex);
+            for (size_t i = startLineIndex; i < currentRuns->size() && i >= 0; direction == AXDirection::Next ? i++ : i--) {
+                // If a text run starts or ends with a newline character, that indicates a paragraph boundary.
+                if (currentRuns->at(i).endsWithLineBreak() && (i != startLineIndex || (direction == AXDirection::Next && currentRuns->runLength(i) < offsetInStartLine))) {
+                    unsigned newlineOffsetConsideringDirection = direction == AXDirection::Next ? cumulativeOffset + currentRuns->runLength(i) - 1 : cumulativeOffset;
+                    return { *currentObject, newlineOffsetConsideringDirection };
+                }
+                if (currentRuns->at(i).startsWithLineBreak() && (i != startLineIndex || direction == AXDirection::Previous)) {
+                    unsigned newlineOffsetConsideringDirection = direction == AXDirection::Next ? 0 : 1;
+                    return { *currentObject, cumulativeOffset + newlineOffsetConsideringDirection };
+                }
+                cumulativeOffset = direction == AXDirection::Next ? cumulativeOffset + currentRuns->runLength(i) : cumulativeOffset - currentRuns->runLength(i);
+            }
+
+            RefPtr previousObject = currentObject;
+            const auto* previousRuns = previousObject->textRuns();
+            currentObject = findObjectWithRuns(*currentObject, direction, stopAtID);
+            currentRuns = currentObject ? currentObject->textRuns() : nullptr;
+
+            // Paragraph boundarys also change based on editability, containing block, and whether we hit a line break.
+            bool isContainingBlockBoundary = currentRuns && previousRuns && currentRuns->containingBlock != previousRuns->containingBlock;
+            bool isEditBoundary = previousObject && currentObject && !!previousObject->editableAncestor() != !!currentObject->editableAncestor();
```

Comment:
We don't need to compute `isEditBoundary` if `isContainingBlockBoundary`. Since the latter does two ancestry walks, maybe we could do this with a comment making it clear why we do it, even if it's a bit verbose:

```cpp
bool isContainingBlockBoundary = ...;
// Don't bother computing isEditBoundary if isContainingBlockBoundary since we only need one or the other below.
bool isEditBoundary = !isContainingBlockBoundary && previousObject && ...;
```